File: client/src/Solvers/RowPoints.ts
import { MatchedWord, Tile } from "../Models/Tile";
import { countWordPoint, countCharPoint } from './CountPoints';

const specials = (currentPoints: number, rowWord: MatchedWord, board: Array<Array<Tile>>) => {
  let points = currentPoints;
  for (let i = 0; i < rowWord.word.length; i++) {
    points = countWordPoint(currentPoints, board[rowWord.row][rowWord.column + i])
  }

  return points
}

const rowPoints = (rowWord: MatchedWord, board: Array<Array<Tile>>): number => {
  let points = 0;
  for (let i = 0; i < rowWord.word.length; i++) {
    points += countCharPoint(board[rowWord.row][rowWord.column + i], rowWord.word[i]);
  }

  return specials(points, rowWord, board);
}

const findRowWords = (board: Array<Array<Tile>>): Array<MatchedWord> => {
  const wordsFound: Array<MatchedWord> = [];

  for (let row = 0; row < board.length; row++) {
    const matchedWords: Array<MatchedWord> = [{
      word: '', points: 0, direction: 'row', row, column: -1, hasNotFinalCharacter: false,
    }];

    for (let column = 0; column < board.length; column++) {
      if (board[row][column].char === '') {
        if (matchedWords[matchedWords.length - 1].word.length > 0) {
          matchedWords.push({ word: '', points: 0, direction: 'row', column: -1, row, hasNotFinalCharacter: false })
        }
      } else {
        if (matchedWords[matchedWords.length - 1].column === -1) {
          matchedWords[matchedWords.length - 1].column = column;
        }

        matchedWords[matchedWords.length - 1].word += board[row][column].char;
        matchedWords[matchedWords.length - 1].hasNotFinalCharacter =
          matchedWords[matchedWords.length - 1].hasNotFinalCharacter || !board[row][column].final;
      }
    }
    wordsFound.push(...matchedWords.filter(matchedWord => matchedWord.word.length > 1 && matchedWord.hasNotFinalCharacter))
  }

  return wordsFound;
}

const countRowPoints = (board: Array<Array<Tile>>) => {
  const wordsFound: Array<MatchedWord> = findRowWords(board)

  return wordsFound.reduce((previous, current) => {
    return previous + rowPoints(current, board)
  }, 0)
}

export {
  countRowPoints,
  findRowWords,
}

File: server/rowPoints.go
package main

func specialsRow(currentPoints int, matchedWord MatchedWord, board *Board) int {
	points := currentPoints
	for i := 0; i < len(matchedWord.word); i++ {
		if matchedWord.column+i >= boardLength {
			break
		}

		points = countWordPoint(currentPoints, board.tiles[matchedWord.row][matchedWord.column+i])
	}

	return points
}

func rowPoints(matchedWord MatchedWord, board *Board) int {
	points := 0
	for i := 0; i < len(matchedWord.word); i++ {
		if matchedWord.column+i >= boardLength {
			break
		}

		points += countCharPoint(board.tiles[matchedWord.row][matchedWord.column+i], string(matchedWord.word[i]))
	}

	return specialsRow(points, matchedWord, board)
}

func findRowWords(board *Board) []MatchedWord {
	wordsFound := []MatchedWord{}

	for row := 0; row < boardLength; row++ {
		matchedWords := []MatchedWord{
			{
				word:                 "",
				points:               0,
				column:               -1,
				row:                  row,
				hasNotFinalCharacter: false,
			},
		}

		for column := 0; column < boardLength; column++ {
			if board.tiles[row][column].c == "" {
				if len(matchedWords[len(matchedWords)-1].word) > 0 {
					matchedWords = append(matchedWords, MatchedWord{
						word:                 "",
						points:               0,
						column:               -1,
						row:                  row,
						hasNotFinalCharacter: false,
					})
				}
			} else {
				if matchedWords[len(matchedWords)-1].column == -1 {
					matchedWords[len(matchedWords)-1].column = column
				}

				matchedWords[len(matchedWords)-1].word += board.tiles[row][column].c
				matchedWords[len(matchedWords)-1].hasNotFinalCharacter =
					matchedWords[len(matchedWords)-1].hasNotFinalCharacter || !board.tiles[row][column].fixed
			}
		}

		// Word has to be at least 2 characters and contain one character that is not final
		for i := 0; i < len(matchedWords); i++ {
			if len(matchedWords[i].word) > 1 && matchedWords[i].row >= 0 && matchedWords[i].hasNotFinalCharacter {
				wordsFound = append(wordsFound, matchedWords[i])
			}
		}
	}

	return wordsFound
}

func countRowPoints(board *Board) int {
	wordsFound := findRowWords(board)
	points := 0

	for i := 0; i < len(wordsFound); i++ {
		points += rowPoints(wordsFound[i], board)
	}

	return points
}
